test/e2e/storage/csi_mock_volume.go

```diff
@@ -897,6 +897,128 @@ var _ = utils.SIGDescribe("CSI mock volume", func() {
 		}
 	})
 
+	ginkgo.Context("CSI NodeUnstage error cases [Slow]", func() {
+		trackedCalls := []string{
+			"NodeStageVolume",
+			"NodeUnstageVolume",
+		}
+
+		// Each test starts two pods in sequence.
+		// The first pod always runs successfully, but NodeUnstage hook can set various error conditions.
+		// The test then checks how NodeStage of the second pod is called.
+		tests := []struct {
+			name          string
+			expectedCalls []csiCall
+
+			// Called for each NodeStageVolume calls, with counter incremented atomically before
+			// the invocation (i.e. first value will be 1) and index of deleted pod (the first pod
+			// has index 1)
+			nodeUnstageHook func(counter, pod int64) error
+		}{
+			{
+				// This is already tested elsewhere, adding simple good case here to test the test framework.
+				name: "should call NodeStage after NodeUnstage success",
+				expectedCalls: []csiCall{
+					{expectedMethod: "NodeStageVolume", expectedError: codes.OK},
+					{expectedMethod: "NodeUnstageVolume", expectedError: codes.OK},
+					{expectedMethod: "NodeStageVolume", expectedError: codes.OK},
+					{expectedMethod: "NodeUnstageVolume", expectedError: codes.OK},
+				},
+			},
+			{
+				name: "two pods: should call NodeStage after previous NodeUnstage final error",
+				expectedCalls: []csiCall{
+					{expectedMethod: "NodeStageVolume", expectedError: codes.OK},
+					{expectedMethod: "NodeUnstageVolume", expectedError: codes.InvalidArgument},
```

Comment:
For InvalidArgument, do we need to mark as uncertain?

Reply:
It's an example of a "final" response, where we can be reasonably sure that e.g. provisioning is not in progress. Not sure if we can apply the same logic to Unstage.